File: Programmers/괄호 변환.py
def solution(p):
    answer = ''

    # step 1
    # step 2
    answer = find_str(p)

    return answer


def find_str(input):
    answer = ""
    if len(input) == 0:
        return answer
    else:
        u, v = separate_uv(input)

        if check_right_str(u) == True:
            answer += u
            answer += find_str(v)


        else:
            temp = "("
            temp += find_str(v)
            temp += ")"
            u = u[1:-1]
            temp += reverse_flag(u)
            answer = temp

    return answer


def separate_uv(input):
    left_count = 0  # (
    right_count = 0  # )

    for i in range(len(input)):
        if input[i] == "(":
            left_count += 1
        else:
            right_count += 1

        if left_count == right_count:
            break;

    return input[:left_count + right_count], input[left_count + right_count:]


def check_right_str(input):
    stack = []
    for str in input:
        if str == "(":
            stack.append(str)
        else:
            if len(stack) == 0:
                return False
            else:
                stack.pop()
    return True


def reverse_flag(input):
    temp = ""
    for flag in input:
        if flag == ")":
            temp += "("
        else:
            temp += ")"

    return temp

File: Programmers JS/괄호 변환.js
function solution(p) {
  var answer = "";
  answer = process(p);
  return answer;
}
function process(string) {
  let split = divide(string);
  let u = split[0];
  let v = split[1];

  if (right(u)) {
    v = v === "" ? "" : process(v);
    return u + v;
  } else {
    let temp = "(";
    temp += v === "" ? "" : process(v);
    temp += ")";
    u = u.substring(1, u.length - 1);
    u = u
      .split("")
      .map((c) => (c === "(" ? ")" : "("))
      .join("");
    temp += u;
    return temp;
  }
}

function divide(string) {
  let left_count = 0;
  let right_count = 0;
  let index = 0;

  while (index < string.length) {
    if (string[index] === "(") {
      left_count++;
    } else {
      right_count++;
    }
    index++;
    if (left_count === right_count) {
      break;
    }
  }
  const u = string.substring(0, index);
  const v = string.substring(index);
  return [u, v];
}

function right(string) {
  if (string[0] === ")") {
    return false;
  }
  let stack = [];
  let arr = string.split("");
  while (arr.length > 0) {
    let pop = arr.shift();
    if (stack.length === 0) {
      stack.push(pop);
    } else if (stack[stack.length - 1] === pop) {
      stack.push(pop);
    } else if (stack[stack.length - 1] !== pop) {
      stack.pop();
    }
  }

  return stack.length === 0 ? true : false;
}
